sys/usb/usbus/usbus.c

```diff
@@ -0,0 +1,334 @@
+    }
+    else {
+        usbus->iface = iface;
+    }
+    iface->idx = idx;
+    return idx;
+}
+
+void usbus_register_event_handler(usbus_t *usbus, usbus_handler_t *handler)
+{
+    /* See note above for reasons against clist.h */
+    usbus_handler_t *last = usbus->handlers;
+
+    if (last) {
+        while (last->next) {
+            last = last->next;
+        }
+        last->next = handler;
+    }
+    else {
+        usbus->handlers = handler;
+    }
+}
+
+int usbus_add_endpoint(usbus_t *usbus, usbus_interface_t *iface,
+                       usbus_endpoint_t *ep, usb_ep_type_t type,
+                       usb_ep_dir_t dir, size_t len)
+{
+    int res = -ENOMEM;
+    usbdev_ep_t *usbdev_ep = usbdev_new_ep(usbus->dev, type, dir, len);
+
+    if (usbdev_ep) {
+        ep->maxpacketsize = usbdev_ep->len;
+        ep->ep = usbdev_ep;
+        ep->next = iface->ep;
+        iface->ep = ep;
+        res = 0;
+    }
+    return res;
+}
+
+static void _signal_handlers(usbus_t *usbus, uint16_t flag,
+                             uint16_t msg)
+{
+    for (usbus_handler_t *handler = usbus->handlers;
+         handler; handler = handler->next) {
+        if (handler->flags & flag) {
+            handler->driver->event_handler(usbus, handler, msg);
+        }
+    }
+}
+
+static void _usbus_init_handlers(usbus_t *usbus)
+{
+    for (usbus_handler_t *handler = usbus->handlers;
+         handler; handler = handler->next) {
+        handler->driver->init(usbus, handler);
+    }
+}
+
+static void *_usbus_thread(void *args)
+{
+    usbus_t *usbus = (usbus_t *)args;
+    usbus_control_handler_t ep0_handler;
+
+    usbus->control = &ep0_handler.handler;
+
+    usbus_control_init(usbus, &ep0_handler);
+
+    usbdev_t *dev = usbus->dev;
+    usbus->pid = sched_active_pid;
+    usbus->addr = 0;
+    usbus->iface = NULL;
+    usbus->str_idx = 1;
+    msg_t msg, msg_queue[_USBUS_MSG_QUEUE_SIZE];
+    DEBUG("usbus: starting thread %i\n", sched_active_pid);
+    /* setup the link-layer's message queue */
+    msg_init_queue(msg_queue, _USBUS_MSG_QUEUE_SIZE);
+    /* register the event callback with the device driver */
+    dev->cb = _event_cb;
+    dev->epcb = _event_ep_cb;
+    /* initialize low-level driver */
+    dev->context = usbus;
+    usbdev_init(dev);
+
+    usbus_add_string_descriptor(usbus, &usbus->config,
+                                USB_CONFIG_CONFIGURATION_STR);
+    usbus_add_string_descriptor(usbus, &usbus->product, USB_CONFIG_PRODUCT_STR);
+    usbus_add_string_descriptor(usbus, &usbus->manuf, USB_CONFIG_MANUF_STR);
+
+    usbus->state = USBUS_STATE_DISCONNECT;
+
+    /* Initialize handlers */
+    _usbus_init_handlers(usbus);
+
+#if (USBUS_AUTO_ATTACH)
+    static const usbopt_enable_t _enable = USBOPT_ENABLE;
+    usbdev_set(dev, USBOPT_ATTACH, &_enable,
+               sizeof(usbopt_enable_t));
+#endif
+
+    while (1) {
+        msg_receive(&msg);
+        /* dispatch USBUS messages */
+        switch (msg.type & 0xFF00) {
```

Comment:
Is the mask really needed ? If yes, please add a note for explaining black magic.

Reply:
refactored out while implementing [the comment](https://github.com/RIOT-OS/RIOT/pull/10916#discussion_r253214624) from @kaspar030